File: runner-contract/src/main/java/dev/vlaisanem/automation/runner/contract/RunnerEvent.java
package dev.vlaisanem.automation.runner.contract;

import java.time.Instant;
import java.util.Objects;

/**
 * Event envelope shared between the JUnit listener, the {@code Steps} API, and the runner service,
 * schema version 1.1. The runner service owns run-level lifecycle events; the listener owns
 * test-level execution events; the {@code Steps} API (main automation suite) owns step-level
 * events. The contract is deliberately framework-agnostic (no serialization annotations), so each
 * side configures its own JSON mapper.
 *
 * @param runOutcome terminal process outcome; required for {@link EventType#RUN_FINISHED} and
 *     absent for every other event type.
 * @param testId JUnit's {@code TestIdentifier.getUniqueId()}; required for test-level and
 *     step-level types, absent for every run-level type ({@link EventType#RUN_QUEUED}, {@link
 *     EventType#RUN_STARTED}, {@link EventType#RUN_FINISHED}).
 * @param testDisplayName {@code TestIdentifier.getDisplayName()}; required for test-level and
 *     step-level types, absent for run-level types.
 * @param stepId opaque identifier for one step within a test, scoped to that test; required for
 *     step-level types, absent for every other event type.
 * @param stepName human-readable step name; required for step-level types, absent for every other
 *     event type.
 * @param detail failure message, skip reason, or {@code null} when not applicable.
 */
public record RunnerEvent(
    String schemaVersion,
    String runId,
    long sequence,
    Instant timestamp,
    EventType type,
    RunOutcome runOutcome,
    String testId,
    String testDisplayName,
    String stepId,
    String stepName,
    String detail) {

  public static final String CURRENT_SCHEMA_VERSION = "1.1";

  public RunnerEvent {
    if (schemaVersion == null || schemaVersion.isBlank()) {
      throw new IllegalArgumentException("schemaVersion must not be blank");
    }
    if (runId == null || runId.isBlank()) {
      throw new IllegalArgumentException("runId must not be blank");
    }
    if (sequence < 1) {
      throw new IllegalArgumentException("sequence must be positive, was " + sequence);
    }
    Objects.requireNonNull(timestamp, "timestamp must not be null");
    Objects.requireNonNull(type, "type must not be null");
    if (type == EventType.RUN_FINISHED) {
      Objects.requireNonNull(runOutcome, "RUN_FINISHED requires a runOutcome");
    } else if (runOutcome != null) {
      throw new IllegalArgumentException(type + " must not carry a runOutcome");
    }
    if (type.isStepLevel()) {
      if (testId == null || testId.isBlank()) {
        throw new IllegalArgumentException(type + " requires a non-blank testId");
      }
      if (testDisplayName == null || testDisplayName.isBlank()) {
        throw new IllegalArgumentException(type + " requires a non-blank testDisplayName");
      }
      if (stepId == null || stepId.isBlank()) {
        throw new IllegalArgumentException(type + " requires a non-blank stepId");
      }
      if (stepName == null || stepName.isBlank()) {
        throw new IllegalArgumentException(type + " requires a non-blank stepName");
      }
    } else if (type.isTestLevel()) {
      if (testId == null || testId.isBlank()) {
        throw new IllegalArgumentException(type + " requires a non-blank testId");
      }
      if (testDisplayName == null || testDisplayName.isBlank()) {
        throw new IllegalArgumentException(type + " requires a non-blank testDisplayName");
      }
      if (stepId != null || stepName != null) {
        throw new IllegalArgumentException(type + " must not carry a step identifier");
      }
    } else {
      if (testId != null || testDisplayName != null) {
        throw new IllegalArgumentException(type + " must not carry a test identifier");
      }
      if (stepId != null || stepName != null) {
        throw new IllegalArgumentException(type + " must not carry a step identifier");
      }
    }
  }

  public static RunnerEvent runQueued(String runId, long sequence, Instant timestamp) {
    return new RunnerEvent(
        CURRENT_SCHEMA_VERSION,
        runId,
        sequence,
        timestamp,
        EventType.RUN_QUEUED,
        null,
        null,
        null,
        null,
        null,
        null);
  }

  public static RunnerEvent runStarted(String runId, long sequence, Instant timestamp) {
    return new RunnerEvent(
        CURRENT_SCHEMA_VERSION,
        runId,
        sequence,
        timestamp,
        EventType.RUN_STARTED,
        null,
        null,
        null,
        null,
        null,
        null);
  }

  public static RunnerEvent runFinished(
      String runId, long sequence, Instant timestamp, RunOutcome runOutcome, String detail) {
    return new RunnerEvent(
        CURRENT_SCHEMA_VERSION,
        runId,
        sequence,
        timestamp,
        EventType.RUN_FINISHED,
        runOutcome,
        null,
        null,
        null,
        null,
        detail);
  }

  public static RunnerEvent testStarted(
      String runId, long sequence, Instant timestamp, String testId, String testDisplayName) {
    return new RunnerEvent(
        CURRENT_SCHEMA_VERSION,
        runId,
        sequence,
        timestamp,
        EventType.TEST_STARTED,
        null,
        testId,
        testDisplayName,
        null,
        null,
        null);
  }

  public static RunnerEvent testPassed(
      String runId, long sequence, Instant timestamp, String testId, String testDisplayName) {
    return new RunnerEvent(
        CURRENT_SCHEMA_VERSION,
        runId,
        sequence,
        timestamp,
        EventType.TEST_PASSED,
        null,
        testId,
        testDisplayName,
        null,
        null,
        null);
  }

  public static RunnerEvent testFailed(
      String runId,
      long sequence,
      Instant timestamp,
      String testId,
      String testDisplayName,
      String failureMessage) {
    return new RunnerEvent(
        CURRENT_SCHEMA_VERSION,
        runId,
        sequence,
        timestamp,
        EventType.TEST_FAILED,
        null,
        testId,
        testDisplayName,
        null,
        null,
        failureMessage);
  }

  public static RunnerEvent testAborted(
      String runId,
      long sequence,
      Instant timestamp,
      String testId,
      String testDisplayName,
      String reason) {
    return new RunnerEvent(
        CURRENT_SCHEMA_VERSION,
        runId,
        sequence,
        timestamp,
        EventType.TEST_ABORTED,
        null,
        testId,
        testDisplayName,
        null,
        null,
        reason);
  }

  public static RunnerEvent testSkipped(
      String runId,
      long sequence,
      Instant timestamp,
      String testId,
      String testDisplayName,
      String reason) {
    return new RunnerEvent(
        CURRENT_SCHEMA_VERSION,
        runId,
        sequence,
        timestamp,
        EventType.TEST_SKIPPED,
        null,
        testId,
        testDisplayName,
        null,
        null,
        reason);
  }

  public static RunnerEvent stepStarted(
      String runId,
      long sequence,
      Instant timestamp,
      String testId,
      String testDisplayName,
      String stepId,
      String stepName) {
    return new RunnerEvent(
        CURRENT_SCHEMA_VERSION,
        runId,
        sequence,
        timestamp,
        EventType.STEP_STARTED,
        null,
        testId,
        testDisplayName,
        stepId,
        stepName,
        null);
  }

  public static RunnerEvent stepPassed(
      String runId,
      long sequence,
      Instant timestamp,
      String testId,
      String testDisplayName,
      String stepId,
      String stepName) {
    return new RunnerEvent(
        CURRENT_SCHEMA_VERSION,
        runId,
        sequence,
        timestamp,
        EventType.STEP_PASSED,
        null,
        testId,
        testDisplayName,
        stepId,
        stepName,
        null);
  }

  public static RunnerEvent stepFailed(
      String runId,
      long sequence,
      Instant timestamp,
      String testId,
      String testDisplayName,
      String stepId,
      String stepName,
      String failureMessage) {
    return new RunnerEvent(
        CURRENT_SCHEMA_VERSION,
        runId,
        sequence,
        timestamp,
        EventType.STEP_FAILED,
        null,
        testId,
        testDisplayName,
        stepId,
        stepName,
        failureMessage);
  }
}

File: runner-dashboard/src/domain/runner-event.ts
import { z } from "zod";

/**
 * Hand-written, not generated: the SSE endpoint is excluded from the OpenAPI document (see
 * `docs/SSE_CONTRACT_V1.md`). Keep in sync with `runner-contract`'s `RunnerEvent` record by hand.
 */
export const RunOutcome = z.enum([
  "SUCCEEDED",
  "FAILED",
  "TIMED_OUT",
  "CANCELLED",
  "ERROR",
]);
export type RunOutcome = z.infer<typeof RunOutcome>;

const runLevelEventTypes = [
  "RUN_QUEUED",
  "RUN_STARTED",
  "RUN_FINISHED",
] as const;
const testLevelEventTypes = [
  "TEST_STARTED",
  "TEST_PASSED",
  "TEST_FAILED",
  "TEST_ABORTED",
  "TEST_SKIPPED",
] as const;
const stepLevelEventTypes = [
  "STEP_STARTED",
  "STEP_PASSED",
  "STEP_FAILED",
] as const;

export const EventType = z.enum([
  ...runLevelEventTypes,
  ...testLevelEventTypes,
  ...stepLevelEventTypes,
]);
export type EventType = z.infer<typeof EventType>;

/** Java's `String.isBlank()` rejects `""` and whitespace-only strings alike; `z.string()` alone accepts both. */
const nonBlankString = z.string().regex(/\S/, "must not be blank");

/**
 * Minimal shape used to check `runId`/`schemaVersion` before full V1 validation (see
 * `run-event-reducer.ts`). Loose about everything else so a future V2 event's unknown shape is
 * classified as an unsupported-schema-version error, not a generic protocol error.
 */
export const RunnerEventEnvelope = z.object({
  schemaVersion: nonBlankString,
  runId: nonBlankString,
});
export type RunnerEventEnvelope = z.infer<typeof RunnerEventEnvelope>;

/**
 * Optional fields are `.optional()`, never nullable - the backend omits the key rather than
 * sending `null`. `.strict()` mirrors the backend's cross-scope-field rejection (e.g. a run-level
 * event can't carry `testId`).
 */
const baseFields = {
  schemaVersion: nonBlankString,
  runId: nonBlankString,
  sequence: z.int().positive(),
  timestamp: z.iso.datetime(),
  detail: z.string().optional(),
};

const RunQueuedEvent = z
  .object({ ...baseFields, type: z.literal("RUN_QUEUED") })
  .strict();
const RunStartedEvent = z
  .object({ ...baseFields, type: z.literal("RUN_STARTED") })
  .strict();
const RunFinishedEvent = z
  .object({
    ...baseFields,
    type: z.literal("RUN_FINISHED"),
    runOutcome: RunOutcome,
  })
  .strict();

const testLevelFields = {
  ...baseFields,
  testId: nonBlankString,
  testDisplayName: nonBlankString,
};

const TestStartedEvent = z
  .object({ ...testLevelFields, type: z.literal("TEST_STARTED") })
  .strict();
const TestPassedEvent = z
  .object({ ...testLevelFields, type: z.literal("TEST_PASSED") })
  .strict();
const TestFailedEvent = z
  .object({ ...testLevelFields, type: z.literal("TEST_FAILED") })
  .strict();
const TestAbortedEvent = z
  .object({ ...testLevelFields, type: z.literal("TEST_ABORTED") })
  .strict();
const TestSkippedEvent = z
  .object({ ...testLevelFields, type: z.literal("TEST_SKIPPED") })
  .strict();

/**
 * Additive over schema 1.0's `RUN_*`/`TEST_*` vocabulary; emitted by the `Steps` API from inside a
 * test method, never by the JUnit listener. A step-free test simply emits none of these.
 */
const stepLevelFields = {
  ...testLevelFields,
  stepId: nonBlankString,
  stepName: nonBlankString,
};

const StepStartedEvent = z
  .object({ ...stepLevelFields, type: z.literal("STEP_STARTED") })
  .strict();
const StepPassedEvent = z
  .object({ ...stepLevelFields, type: z.literal("STEP_PASSED") })
  .strict();
const StepFailedEvent = z
  .object({ ...stepLevelFields, type: z.literal("STEP_FAILED") })
  .strict();

export const RunnerEvent = z.discriminatedUnion("type", [
  RunQueuedEvent,
  RunStartedEvent,
  RunFinishedEvent,
  TestStartedEvent,
  TestPassedEvent,
  TestFailedEvent,
  TestAbortedEvent,
  TestSkippedEvent,
  StepStartedEvent,
  StepPassedEvent,
  StepFailedEvent,
]);
export type RunnerEvent = z.infer<typeof RunnerEvent>;

export type RunLevelEventType = (typeof runLevelEventTypes)[number];
export type TestLevelEventType = (typeof testLevelEventTypes)[number];
export type StepLevelEventType = (typeof stepLevelEventTypes)[number];

export function isTestLevelEvent(
  event: RunnerEvent,
): event is Extract<RunnerEvent, { type: TestLevelEventType }> {
  return (testLevelEventTypes as readonly string[]).includes(event.type);
}

export function isStepLevelEvent(
  event: RunnerEvent,
): event is Extract<RunnerEvent, { type: StepLevelEventType }> {
  return (stepLevelEventTypes as readonly string[]).includes(event.type);
}

/** The schema version this frontend build was written against - see `RunnerEvent.CURRENT_SCHEMA_VERSION` in `runner-contract`. */
export const CURRENT_SCHEMA_VERSION = "1.1";
